rfc: add RFC for Talent Plan Course (#140)
pingcap/community: rfc/2019-12-11-Talent-Plan-Courses.md

# Proposal: Talent Plan Courses

- Author(s): [Ling Jin](https://github.com/JinLingChristoher)
- Last updated: 2019-12-11
- Discussion at: [Incubating Program: Talent Plan Courses](https://github.com/pingcap/community/issues/130)

> There are plenty of people interested in TiDB but lacking background knowledge, so we propose to create a series of courses to help those people enter the field infrastructure systems and TiDB ecosystem. Course participants should have developed good knowledge and skills in system programming after finishing course materials.

Reviewer: ```suggestion
There are plenty of people interested in TiDB but lacking background knowledge, so we propose to create a series of courses to help those people enter the field of infrastructure systems and TiDB ecosystem. Course participants should have developed good knowledge and skills in system programming after learning the series of courses.
```

> There are plenty of people interested in TiDB but lacking background knowledge, so we propose to create a series of courses to help those people enter the field of infrastructure systems and TiDB ecosystem. Course participants should have developed good knowledge and skills in system programming after learning the series of courses.

Reviewer 2: ```suggestion
There are plenty of people interested in TiDB but lacking background knowledge, so we propose to create a series of courses to help those people enter the field of infrastructure systems and TiDB ecosystem. After learning the series of courses, course participants should be able to develop good knowledge and skills in system programming.
```

## Course design

> ![course map](./course-ex.png)

Reviewer 3: 1. the course graph needs to be updated.
2. please put `course-ex.png` to the `media` folder. you can name it to `rfc-talent-plan-courses.png`

> The picture above is the design of the course system. This series of courses include but not limited to infrastructure systems, the programming language to be used, open-source collaboration. The development of a distributed database system course is in progress, and more courses will be created in the future.

Reviewer 2: ```suggestion
The picture above is the design of the course system. This series of courses include but not limited to infrastructure systems, the programming language to be used, and open-source collaboration. The development of a distributed database system course is in progress, and more courses will be provided in the future.
```

> The above is the design of the course system. This series of courses include but not limited to the principles of computer system design, the programming language to be used, open-source cooperation, and more. The development of a distributed database system course is in progress, and more courses will be created in the future.

Reviewer: ```suggestion
The picture above is the design of the course system. This series of courses include but not limited to infrastructure systems, the programming language to be used, open-source collaboration. The development of a distributed database system course is in progress, and more courses will be created in the future.
```

> The above is a introduction of the course system. This series of courses include but not limited to the principle of computer system design, but also programming language, open source cooperation, and more. The development of a distributed database system course is on the way, more courses will be created in the future. We hope those courses can make benefit to everyone.

Reviewer 2: ```suggestion
The above is the design of the course system. This series of courses include but not limited to the principles of computer system design, the programming language to be used, open-source cooperation, and more. The development of a distributed database system course is in progress, and more courses will be created in the future.
```

## Rationale

> There are many open study resources online, our courses can borrow ideas from world-class universities and MOOCs as well. The courses will be practice-concentrate, with well-designed programming assignments as the core, Reading materials are also included.

Reviewer 2: ```suggestion
There are many open study resources online.  Our courses can borrow ideas from world-class universities and MOOCs as well. The courses will be practice-oriented, with well-designed programming assignments as the core. Reading materials are also included.
```

> There are many open study resources online, our courses can borrow ideas from world-class universities and from MOOCs as well. The courses will be practice-concentrate, with well-designed programming assignments as the core. Reading materials including slides, notes is also irreplaceable.

Reviewer 2: > Reading materials include slides, notes is also irreplaceable.

Can you explain this? I don't get it.

Author: The slide might be the most widely used form to represent teach contents, but the note is also important and should not be ignored.

> There are many open study resources online, our courses can borrow ideas from world top universities also MOOCs. The courses would be practice-concentrated, well-designed programming assignments are the core, reading materials including slides, notes is also irreplaceable.

Reviewer 2: ```suggestion
There are many open study resources online, our courses can borrow ideas from world-class universities and from MOOCs as well. The courses will be practice-concentrate, with well-designed programming assignments as the core. Reading materials including slides, notes is also irreplaceable.
```

## Compatibility

N/A

## Implementation

> To implement it, we need the following main steps:

Reviewer 2: ```suggestion
The implementation of the course requires the following 3 main steps:  
```

- Discuss the subject to be included in the course system.
- Divide the subject topics, and assign them to course developers.

> - Create the programming assignment framework and the corresponding reading materials.

Reviewer 2: ```suggestion
- Create the programming assignment framework and prepare the corresponding reading materials.
```

# Proposal: Talent Plan Courses

- Author(s): [Ling Jin](https://github.com/JinLingChristoher)
- Last updated: 2019-12-11
- Discussion at: [Incubating Program: Talent Plan Courses](https://github.com/pingcap/community/issues/130)

> 

Reviewer 3: 1 empty line is enough.

# Proposal: Talent Plan Courses

- Author(s): [Ling Jin](https://github.com/JinLingChristoher)
- Last updated: 2019-12-11
- Discussion at: [Incubating Program: Talent Plan Courses](https://github.com/pingcap/community/issues/130)

> ## Abstract

Reviewer 2: ```suggestion
```

There are plenty of people interested in TiDB but lacking background knowledge, so we propose to create a series of courses to help those people enter the field infrastructure systems and TiDB ecosystem. Course participants should have developed good knowledge and skills in system programming after finishing course materials.

> ## Proposal

Reviewer 2: ```suggestion
## Course design
```

# Proposal: Talent Plan Courses

- Author(s): [Ling Jin](https://github.com/JinLingChristoher)
- Last updated: 2020-01-05
- Discussion at: [Incubating Program: Talent Plan Courses](https://github.com/pingcap/community/issues/130)

## Background

Talent Plan is a study project developed by PingCAP, which aims to help people get in touch with TiDB and TiKV. At the moment, only two study roadmaps were provided, and the content is not well-designed. So we propose Talent Plan Courses, a new version of the study project that contains more topics and elaborate contents.

> ## Course design

Reviewer: ```suggestion
## Course Framework
```

> Since there are plenty of people interested in TiDB but lack background knowledge, so we propose to create a series of computer system courses to help those people enter the field of the TiDB ecosystem, also computer system infrastructure. Course participants should have developed good knowledge and skills in system programming after consumed course materials.

Reviewer 3: ```suggestion
Since there are plenty of people interested in TiDB but lack background knowledge, so we propose to create a series of courses to help those people enter the field of infrastructure system and TiDB ecosystem. Course participants should have developed good knowledge and skills in system programming after consumed course materials.
```

> The campus is always an important member of the TiDB Community, and we hope more young blood from universities or other fields can embrace the TiDB ecosystem, so we think a series of courses about computer systems is necessary for helping more people get in touch with TiDB.

Reviewer 3: no need to highlight `campus` and `universities`. It's a series of courses that everyone can participant in. For example, high school students, graduates and undergraduates, etc.

## Proposal

> This series of courses concentrate on computer system related fields, include but not limited to the database system, distributed system. The development of a distributed database system course is on the way, more course will be created in the future. We hope those courses will push the campus computer system education forward.

Reviewer 3: s/computer system/computer science/?

Reviewer 3: Could you use a graph to explain the whole architecture of this courses series?

> The picture above is the design of Talent Plan Courses which is made up of multiple topics. Each topic contains multiple courses and has a study goal. In detail, for open source collaboration, we hope participants acquired basic knowledge about open source community operation and knows the basic usage of related tools. For Infrastructure systems, multiple secondary topics will be contained, programming assignments and related reading materials will be provided. Furthermore, Programming languages will be provided as crash study materials for people without related experience, after learning the series of courses, participants should have a good knowledge and skills in infrastructure systems or other related fields. Deep dive into the TiDB ecosystem also is an indispensable part to help people get to know more about TiDB related projects.

Reviewer: ```suggestion
The picture above shows the latest framework of Talent Plan Courses. In addition to TiDB related learning materials, lots of explanations on the basic knowledge of programming language and Infrastructure systems will be provided. Courses can be combined according to students' interests and goals. For example:

- Programming language enthusiasts can learn Golang, which is widely used in well-known enterprises, and the Rust tutorial designed by Rust author, Brian Anderson

- Storage enthusiasts can learn Golang and distributed key-value database internal

- Database enthusiasts can implement a small distributed database by learning Go language, distributed relational database course, and Deep Dive into TiDB.

- Open source enthusiasts can better understand open source software, open source community operation by learning courses related to open source collaboration and practicing in TiDB open source community 

In addition to the learning path planned by the Talent Plan, you can create your own learning path according to your own needs to create paths independent of the learning system of Talent Plan, such as:

- Those who are interested in Web development can learn Golang first through Talent Plan courses and then learn the basics of Golang Web development
- Those interested in Open Source community operations can follow up the Open Source Collaboration categories with additional books (e.g., The Art of Community, Cathedral and Bazaars, etc.)
```

## Rationale

There are many open study resources online, our courses can borrow ideas from world top universities also MOOCs. The content of courses would focus on practice, well-designed programming assignments are the core, reading materials including slides, notes is also irreplaceable.

## Compatibility

> TiDB will not be affected by this project.

Reviewer 3: You can simply write a `N/A`, which is short for `not applicable`

> There are many open study resources online. Our courses can borrow ideas from world-class universities and MOOCs as well. The courses will be practice-oriented, with well-designed programming assignments as the core. Reading materials are also included.

Reviewer: I don't think we have the right to "borrow ideas from world-class universities and MOOCs". What we are doing is gathering the open learning materials together to make the learning path easier and providing some assignments to help students check their learning outcomes.

## Implementation

The implementation of the course requires the following 3 main steps:

> - Discuss the subject to be included in the course system.

Reviewer: ```suggestion
- Discuss the course categories to be included in the new version.
```

> - Divide the subject topics, and assign them to course developers.

Reviewer: ```suggestion
- Make agreements for the syllabus for each course
```

> - Create the programming assignment framework and prepare the corresponding reading materials.

Reviewer: ```suggestion
- Fill in learning materials and assignments.
- Optimize assignment submission and evaluation.

```

- Programming Language

  For anyone who wants to participate in an open-source project, knows the project's development programming language is the most basic requirement. We will deliver some courses about programming language to help participant learns it quickly.

  - Golang
    The Go programming language is heavily used in industry, also the TiDB ecosystem. We advise everyone to learn a bit about it:

  - [A Tour Of Golang](https://tour.golang.org/) is recommanded since it's a great official tutorial for the newbie to learn the basic language features.
  - [The Go Programming Language](http://www.gopl.io/) is recommended as a reference book for people want to learn the language in detail.
  - [Talent Plan MapReduce](https://github.com/pingcap/talent-plan/tree/master/tidb/mapreduce) will be a great tool for participants to judge their learning outcomes.

  - Rust
    The Rust programming language is becoming more and more popular since it has many advanced features. For people interested in TiKV which is based on rust, we recommend the following materials:

>   - [Learn Rust](https://www.rust-lang.org/learn) is official study materials, it contains a book, an interactive practice assignment, and some code examples. We believe anyone interested in the Rust programming language can benefit from it.

Reviewer 2: ```suggestion
  - [Learn Rust](https://www.rust-lang.org/learn) - Official Rust tutorials that include a book, an interactive practice assignment, and some code examples. We believe anyone interested in the Rust programming language can benefit from it.
```

>   - [Practical Networked Application in Rust](https://github.com/pingcap/talent-plan/tree/master/rust) is a programming assignment framework develop by [Brian Anderson](https://github.com/brson), who was a core team member of the rust community. We strongly suggest taking this assignment to consolidate rust programming skills.

Reviewer 2: ```suggestion
  - [Practical Networked Application in Rust](https://github.com/pingcap/talent-plan/tree/master/rust) - A series of projects that incrementally develop a single Rust project from the ground up into a high-performance, networked, parallel and asynchronous key/value store. Along the way, various real-world and practical Rust development subject matter are explored and discussed. This course is developed by [Brian Anderson](https://github.com/brson), who was one of the core writers of the Rust language.
```

Reviewer 2: A suggestion - maybe you should just copy the original introduction for each item on their specific website, instead of writing all of them again on your own.

>   Other programming languages like C++, Java, Python also in the works, as the development of this course system, more concrete contents will be appended.

Reviewer 2: ```suggestion
  Materials for other related programming languages like C++, Java, Python are in development.
```

- Infrastructure System

>   - Distributed Relational Database Internals is a course made up of many reading materials and programming assignments. A broad range of topics such as relational algebra, parser, runtime, optimizer, transaction, DDL and more are covered in readings. TinySQL is a simplified version of TiDB, used as the programming assignment framework, which let participants can verify their learning outcomes.

Reviewer 2: ```suggestion
  - Distributed Relational Database Internals is a course made up of many reading materials and programming assignments. A broad range of topics such as relational algebra, parser, runtime, optimizer, transaction, DDL and more are covered in readings. As the programming assignment framework, TinySQL, a simplified version of TiDB, is used for course participants to verify their learning outcomes.
```

>   - Distributed Key-Value Database Internals is a course have the same design structure as the one mentioned above, but about distributed key-value database system. The contents of readings will be like storage engine, data partitioning, replication, consistency, scalability, balance, distributed transactions and more. For programming assignment, TinyKV is a simplified version of TiKV, will be provided to participants to judge their learning outcomes.

Reviewer 2: ```suggestion
  - Distributed Key-Value Database Internals is a course about distributed key-value database systems. The reading materials cover storage engine, data partitioning, replication, consistency, scalability, balance, distributed transactions and more. For programming assignment, TinyKV, a simplified version of TiKV, will be provided to course participants to verify their learning outcomes.
```

>   As the development of the course, more concrete content will be proposed in the future. Topics like cloud-native database system and Machine Learning also in the plan.

Reviewer 2: ```suggestion
   Topics such as cloud-native database systems and Machine Learning are also planned.
```

- Deep Dive Series Readings

>   For anyone who wants to get to know more about TiDB related projects, they would benefit a lot by reading Deep Dive into TiDB or Deep Dive into TiKV, which covers related projects in detail. It's worth mentioning that Deep Dive into Cloud TiDB is also in the works. We highly recommend reading these materials to help you be familiar with the TiDB ecosystem.

Reviewer 2: ```suggestion
  For anyone who wants to get to know more about TiDB related projects, they would benefit a lot by reading Deep Dive series articles, including into TiDB and Deep Dive into TiKV. It's worth mentioning that Deep Dive into Cloud TiDB is also in development. We highly recommend reading these materials to help you get familiar with the inner world of the TiDB ecosystem.
```

> For participants, they can create their learning path according to individuals conditions. As the development of courses, more topics, and the detailed course contents would be proposed in the future.

Reviewer 2: ```suggestion
Course participants can choose their own learning path based on personal conditions and learning goals. There will be more topics and courses to be proposed in the future.
```

## Rationale

> There are many open study resources online, we can gather those materials together to make the learning path easier. Furthermore, the courses will be practice-oriented, elaborately designed programming assignments as the core, which can help participants to check their learning outcomes.

Reviewer 2: Suggest removing this section. It does not say anything that's useful or new.

# Proposal: Talent Plan Courses

- Author(s): [Ling Jin](https://github.com/JinLingChristoher)
- Last updated: 2020-01-13
- Discussion at: [Incubating Program: Talent Plan Courses](https://github.com/pingcap/community/issues/130)

## Background

> Talent Plan is a study project developed by PingCAP for training and/or evaluating students, new employees, and new contributors to [TiDB] and [TiKV]. At the moment, only two study roadmaps are provided. This proposal provides Talent Plan Courses, a new version of the study project with more enriched topics and course designs.

Reviewer: ```suggestion
Talent Plan is a learning Plan initiated by PingCAP. The purpose is to help those who like open source, infrastructure and database to gain more knowledge and improve their abilities. At the moment, only two study roadmaps are provided. This proposal provides Talent Plan Courses, a new version of the study project with more enriched topics and course designs.
```

Reviewer: > Talent Plan is a study project developed by PingCAP for training and/or evaluating...

I  don't think training and evaluating is the purpose of Talent Plan

[tidb]: https://github.com/pingcap/tidb
[tikv]: https://github.com/tikv/tikv

## Course Framework

![course map](../media/rfc-talent-plan-courses.png)

The picture above shows the latest framework of Talent Plan Courses. In addition to TiDB related learning materials, lots of other sections are provided as well. Following are the details:

- Open Source Cooperation

  Open source spirit is one of the cores of the TiDB Community. At current, multiple courses regarding this subject are on the list:

  - Introduction to OSS

    A fundamental course about open source and its history.

  - How to use Git&Github

    Git is a free and open source distributed version control system designed to handle everything from small to very large projects with speed and efficiency. Github is a popular website based on Git where people around the world collaborate in open source projects. After taking this course, participants should know the usages of Git and Github.

  - Open Source Practice in TiDB

    TiDB ecosystem is made up of multiple open-source projects on Github. For people who want to contribute to TiDB, open-source practice in TiDB would be helpful for them.

- Programming Language

>   To contribute to the code base of an open-source project, knowing the programming language is a basic requirement. We provide the related programming language courses to help participant learns quickly:

Reviewer: ```suggestion
  To contribute to the code base of an open-source project, knowing the programming language is a basic requirement. We provide the related programming language courses to help participants learn quickly:
```

  - Golang

>     The Go programming language is heavily used in the industry, and also in the TiDB ecosystem. If you haven't learned about it, the following are good places to start:

Reviewer: ```suggestion
    The Go programming language is heavily used in the industry, as well as in the TiDB ecosystem. If you haven't learned about it, the following are good places to start:
```
